tests/unit/sidecars/secrets_fetcher_tests.py

```diff
@@ -0,0 +1,220 @@
+
+
+@dataclasses.dataclass
+class FakeBadVaultClient(secrets_fetcher.VaultClient):
+    token_expiration: datetime.datetime
+    token: str = "token-bad"
+
+    # @typing.override  # TODO: Added in version 3.12.
+    def get_secret(self, secret_name: str) -> tuple[typing.Any, datetime.datetime]:
+        """Return a secret value that is not JSON serializable."""
+        s = self.token_expiration
+        return s, self.token_expiration + datetime.timedelta(seconds=30)
+
+
+class Tests(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        spec = {
+            "vault": {
+                "url": config.DefaultFromEnv(config.String, "BASEPLATE_DEFAULT_VAULT_URL"),
+                "role": config.String,
+                "auth_type": config.Optional(
+                    config.OneOf(**secrets_fetcher.VaultClientFactory.auth_types()),
+                    default=secrets_fetcher.VaultClientFactory.auth_types()["aws"],
+                ),
+                "mount_point": config.DefaultFromEnv(
+                    config.String, "BASEPLATE_VAULT_MOUNT_POINT", fallback="aws-ec2"
+                ),
+            },
+            "output": {
+                "path": config.Optional(config.String, default="/var/local/secrets.json"),
+                "owner": config.Optional(config.UnixUser, default=0),
+                "group": config.Optional(config.UnixGroup, default=0),
+                "mode": config.Optional(config.Integer(base=8), default=0o400),  # type: ignore
+            },
+            "secrets": config.Optional(config.TupleOf(config.String), default=[]),
+            "callback": config.Optional(config.String),
+        }
+
+        parser = configparser.RawConfigParser()
+        with io.StringIO(configini) as f:
+            parser.read_file(f)
+        fetcher_config = dict(parser.items("secret-fetcher"))
+
+        cls.cfg = config.parse_config(fetcher_config, spec)
+
+    def setUp(self):
+        self.setUpPyfakefs()
+        self.fake_fs().create_file("/var/local/secrets.json", contents="initial contents")
+
+        cfg = self.cfg
+        now = datetime.datetime.now(UTC)
+        with unittest.mock.patch("baseplate.sidecars.secrets_fetcher.VaultClientFactory") as mock:
+            instance = mock.return_value
+            instance.get_client.return_value = FakeVaultClient(token_expiration=now)
+            f = secrets_fetcher.VaultClientFactory(
+                cfg.vault.url, cfg.vault.role, cfg.vault.auth_type, cfg.vault.mount_point
+            )
+            secrets_fetcher.fetch_secrets(cfg, f)
+
+    def test_is_file(self):
+        p = pathlib.Path("/var/local/secrets.json")
+        self.assertTrue(p.is_file())
+
+    def test_sets_owner(self):
+        p = pathlib.Path("/var/local/secrets.json")
+        self.assertEqual(p.owner(), whoami)
+
+    def test_sets_group(self):
+        p = pathlib.Path("/var/local/secrets.json")
+        self.assertEqual(p.group(), whoami)
```

Comment:
This seems a bit odd since the username isn't inherently also a group and it's fairly common for there not to be a group with the same name as your username. For example user `nobody` typically has primary group `nogroup`, users on macOS usually have primary group `staff`, etc.

Is this working because of the in-memory filesystem mock or for some other reason?

Reply:
This is because locally I was using a devcontainer under user `vscode`, and in GH CI the user is `root`.

Reply:
Maybe replacing with `grp.getgrgid(os.getegid()).gr_name` would work and also produce the right group ID outside of tests?

Reply:
The docs use `www-data` but that wasn't an easy option in either context. I wasn't concerned much about the specific user, but did want to try and capture the implementation behavior which does change the user, group, and permissions (though this last bit is not under test in this PR).